### DISCUSSION TITLE
OAuthAccountNotLinked

### DISCUSSION THREAD

**Discussion by vorcigernix** [nextauthjs/next-auth#1601] [Help]
**Your question**
Any idea why I am consistently getting OAuthAccountNotLinked with only one provider?

**What are you trying to do**
This is basically a copy of this issue #519 - only differences are that I am using google provider in my own project. I didn't even changed the callback url or anything like that. I did changed jwt setting mentioned as a remedy in bug #519 but it didn't helped.

**Reproduction**
I actually don't know what is causing the problem. There is no error in debug mode and all information seems to be okay to my eye. Creating a new user at the same time works fine. 

**Feedback**
*Documentation refers to searching through [online documentation](https://next-auth.js.org), code comments and issue history. The example project refers to [next-auth-example](https://github.com/iaincollins/next-auth-example).*

* [X] Found the documentation helpful
* [ ] Found documentation but was incomplete
* [ ] Could not find relevant documentation
* [X] Found the example project helpful
* [ ] Did not find the example project helpful

**Comment by JavLG** [comment]
Im dealing with the exact same issue. Help needed!

**Reply by znwhite** [reply]
Same here. My Google signin attempt send this answer on my live site (not locally) as soon as I switched out my `database` option with an `adapter` with a custom `User` schema.

**Comment by vorcigernix** [comment]
Cool, so it could be related to the custom user schema. Still not sure how
to fix it, but if I find a time, I'll replicate the issue with a sample
repository and open this as a bug again (or you can).

**Comment by aaron5670** [comment]
Dealing with same error too... :(

**Comment by ahmedelq** [comment]
Having the same error with Github's OAuth

**Comment by balazsorban44** [comment]
Instead of adding these little useful comments, please someone create a proper reproduction, and open a bug report, otherwise we won't be able to look into this problem.

**Reply by balazsorban44** [reply]
Didn't mean to be rude. Maintaining an open source project is a lot of effort and very time consuming, and people don't often relize when an issue is reported in a form like above, we have as little to go on to help as the original poster. Adding comments like "same" is not helpful to anyone. We the maintainers don't get a better understanding of the nature of the problem, it can even be harmful if it obscures a potentially useful message with actual contextual information. I often mark these comments as spam or off-topic to help other fellow maintainers.

You must help us so we can help you. In OSS, both sides supposed to engage, this is how our community can thrive.

**Reply by lienista** [reply]
I'd like to make a comment as food for thought, rather than a hard request. I wanted to make a comment about account linking - it should be up to the developer to decide whether it happens automatically or not. I don't really see much of a security issue if the user is able to log into Google and Facebook previously. We are creating a scenario where it's making things difficult for BOTH the user and the developer. If the user can log into both accounts, why shouldn't we be doing this link automatically. 

If I were a user of a website, I would fault the site developers for making bad software if they didn't perform this link. If the site is able to automatically pull in my existing data and making life easy for me the user, I would be amazed. I think most of us would have to implement this link manually, and I would rather be part of `next-auth`, rather than having to do it myself every time. I don't see it as a security issue at all, and would rather have it already implemented for me than not. If it's security issue, let the developers decide whether to take that risk

**Comment by and-why** [comment]
I was having this error, and it took me a while, but I actually found that I was somehow signed in. The useSession should have stopped the sign in buttons showing, but something wasn't aligning.

I simply went to http://localhost:3000/api/auth/signout and it allowed me to sign out and then I could sign in again. This might be the issue others are facing, so thought I'd post here.

**Reply by balazsorban44** [reply]
do you have any reproduction/code verifying this?

**Reply by and-why** [reply]
I've tried to reproduce this but couldn't unfortunately, it was weird that I could still use the signout route when the sign in process shouldn't be visible on my front end. Hopefully if someone else has this issue they can determine what caused it. It was in the process of migrating from Apollo/Mongodb/Mongoose to Prisma/PostgreSQL so might be something weird in that? Sorry.

**Comment by joematune** [comment]
For me, my [db model](https://next-auth.js.org/adapters/dynamodb#schema) was incorrect. I figured it out by calling all of the adapter's methods, i.e.

```ts
const adapter = DynamoDBAdapter(client, {  tableName: "next-auth" });

const userById = await adapter.getUser(userId);
const updatedUser = await adapter.updateUser(user);
const userByEmail = await adapter.getUserByEmail(email); // I was getting `null` here
// ...
```
The `getUserByEmail` method was failing and so next-auth attempted to create a new record with the same email ~ resulting in this error.

**Comment by IRediTOTO** [comment]
I get this error too. Can we let users login then in account setting they can link social account with their email? Ex:
```
linkAccount(userId)  // Only can use when user login
```

**Comment by Decad** [comment]
Had this issue as well, turns out my database had the wrong data type for the token columns.

Only found this output when debug mode was turned on:
```
  17 updateUser: ({ id, ...data }) => p.user.update({ where: { id }, data }),
  18 deleteUser: (id) => p.user.delete({ where: { id } }),
→ 19 linkAccount: (data) => p.account.create(
  The provided value for the column is too long for the column's type. Column: for
```

This results in a User being created but the linked account was never created. The first time this happens the error is `Callback` any attempts to signin after that result in `OAuthAccountNotLinked` and the user is stuck in this state.

**Comment by vyber7** [comment]
I had this problem too. I only have GitHub and email provider, and I'm using Prisma with MySQL provider. So to fix this I've added one more field in the Account model in prisma file `refresh_token_expires_in` of type `Int?`, and it worked. Everything else in the file is the same as in the [documentation.](https://next-auth.js.org/adapters/prisma)

![Screenshot from 2022-09-08 12-15-02](https://user-images.githubusercontent.com/29589616/189386461-77493cfa-322a-4f87-aa8e-5e4da61aed25.png)

I'm not sure about other OAuth providers, but GitHub expects this field to be there, otherwise I get an error:

![Screenshot from 2022-09-10 13-48-15](https://user-images.githubusercontent.com/29589616/189496133-8a30a785-2fb7-4694-a2e8-afe2933335af.png)

I hope this was helpful.

**Reply by nbouvrette** [reply]
This is odd, I don't see any mention of `refresh_token_expires_in` in the official doc? https://next-auth.js.org/adapters/models

I also couldn't find much while searching the code for that column - the only place it seems used is on the GitHub adapter: https://github.com/nextauthjs/next-auth/blob/d9a2df3a3d202a355e9d751184e64e14f07f8b96/docs/docs/reference/05-oauth-providers/github.md

**Reply by jasongitmail** [reply]
Was it necessary to update your database adapter in any way too?

**Reply by vyber7** [reply]
No, it wasn't. Just added the field. It is necessary for github provider to work.

**Reply by peiris** [reply]
Thanks @vyber7 for the tip, Looking at your code I was missing `@db.Text` for refresh_token, access_token, and id_token.

<img width="663" alt="Screenshot 2023-11-09 at 10 41 27 PM" src="https://github.com/nextauthjs/next-auth/assets/24228836/4e6f5cbc-58ba-4790-bc23-b0601446adfa">

**Comment by hodgerpodger** [comment]
I encountered this error when trying to setup (github oauth, dynamodb-adapter) combo.  The cause and ultimate resolution of my problem was a dynamodb table schema definition error on my part (global secondary index projection type was set to "KEYS_ONLY" when it should have been "ALL").

**Comment by domharrington** [comment]
I documented how I solved this issue here: https://github.com/nextauthjs/next-auth/issues/519#issuecomment-1421065656. Hope it helps someone!

**Comment by tcodes** [comment]
Hay its simple error. YOu can fix it in very easy way.
1) Go to your Atlas Mongo db collection, delete data inside account and users.
2) Run dev again in terminal 'npm run dev'
thats it. Thats error is fixed for you!
